### ISSUE TITLE
Exceptions thrown in Describe/Context BeforeAll or BeforeEach are undiagnosable.

### ISSUE THREAD

**Issue by dantraMSFT** [pester/Pester#1101]
## 1. Provide a general summary of the issue in the Title above

When a BeforeAll or BeforeEach block throws an exception, Pester error output doesn't include any information about the failure. Instead, a secondary exception is thrown due to the call  to AddTestResult failing.

For example:

[-] Error occurred in Context block 3.4s
  ValidationMetadataException: The attribute cannot be added because variable Result with value  would no longer be valid.
  MethodInvocationException: Exception calling "AddTestResult" with "8" argument(s): "The attribute cannot be added because variable Result with value  would no longer be valid."

The same behavior occurs in both Describe and Context Before* blocks with the only difference being the block type (Context or Describe).

When this occurs, the only solution I've found is to edit the catch blocks in It.ps1 and Describe.ps1 and Write-Host the original failure details.  Neither the console output nor the test result file contain this inforamation.

## 2. Describe Your Environment

Operating System, Pester version and PowerShell version:

```powershell
Pester version     : 4.4.0 /home/psrptest/.local/share/powershell/Modules/Pester/4.4.0/Pester.psm1
PowerShell version : 6.1.0-rc.1
OS version         : Unix 4.15.0.1022
```

## 3. Expected Behavior

If you're describing a bug, tell us what should happen.

The error details ($_) from the catch block should be reported and/or logged.

## 4.Current Behavior

The subsequent call to AddTestResult itself throws and this is the information reported; as indicated above.

```powershell
Context 'Verify NTLM PSSession' {
    BeforeEach {
        # If this throws, a ValidationMetadataException is thrown and reported instead of the exception 
        # caused by the this call.
        $session = New-PSSession -ComputerName $TestVariables.HostName -Credential $TestVariables.Credential -Authentication Negotiate
    }

    It 'Verifies New-PSSession -Authentication Negotiate connects with no errors' -Skip:$Skip {
            $session.ComputerName | Should -BeExactly $TestVariables.HostName
            $session.Transport | Should -BeExactly 'WSMan'
        }
```

## 5. Possible Solution

Have a solution in mind?

https://github.com/pester/Pester/wiki/Contributing-to-Pester has detailed instructions on how to contribute.

## 6. Context

How has this issue affected you? What are you trying to accomplish?

**Comment by nohwnd** [comment]
Thanks for reporting will have a look on this.

**Comment by bergmeister** [comment]
@nohwnd I see these errors now as well after the last update of AppVeyor's `Ubuntu` image where it clearly shows it is a Pester issue because I do nothing in the `Describe` block. For some reason this happens only in the `Ubuntu` image of AppVeyor (not Windows) and I cannot repro this issue locally using Ubuntu 18 in WSL. Let me know if you need more details (like build/code links)
![image](https://user-images.githubusercontent.com/9250262/49888549-30c9fb80-fe37-11e8-9642-940b97bd059b.png)
![image](https://user-images.githubusercontent.com/9250262/49888600-4e976080-fe37-11e8-9169-aca4c955d293.png)

**Comment by nohwnd** [comment]
@bergmeister all our linux an macos builds run fine on Travis, and I am running pester locally in docker just fine as well. Could you share your `appveyor.yml` so I can try it myself? I would actually find it useful to setup Linux build for Assert, two birds one scone :)

**Comment by bergmeister** [comment]
@nohwnd Ok, I opened a few PRs for your Assert module including Linux in AppVeyor. Today, I could reproduce the CI failure locally. I will point you to a simplified repro branch tomorrow. So far I can only say that it only happens on Linux after the update to 6.1 and if the code in question is executed outside of Pester, it is fine, so I suspect a subtle integration bug between PowerShell and Pester.

**Comment by bergmeister** [comment]
@nohwnd Ok, I opened a few PRs for your Assert module including Linux in AppVeyor.
Today, I could reproduce the CI failure locally the first time, it happens only in Ubuntu 18 but not Ubuntu 16 when using WSL for example. I have a simplified repro branch [here](https://github.com/bergmeister/PSScriptAnalyzer/tree/reproPesterBug) where the [appveyor.yml](https://github.com/bergmeister/PSScriptAnalyzer/blob/reproPesterBug/appveyor.yml) just calls `Invoke-Pester ./Tests/Documentation/RuleDocumentation.tests.ps1` in the PSScriptAnalyzer repo (you can take the file from the development branch as well for that). So far I can only say that it only happens on Linux after the minor update to 6.1 and if the code in question is executed outside of Pester, it is fine, so I suspect a subtle integration bug between PowerShell and Pester.

**Comment by nohwnd** [comment]
The errors are most likely filtered out by https://github.com/pester/Pester/blob/master/Functions/Output.ps1#L475 if you get rid of the pipeline it won't limit the output and will print the internals of Pester as well. Could you try that first to see if it gives relevant info please?

**Comment by bergmeister** [comment]
@nohwnd  OK, thanks, this is what I get now as an output:
```
   [-] Error occurred in Describe block 156ms
      ValidationMetadataException: The attribute cannot be added because variable Result with value  would no longer be valid.
      MethodInvocationException: Exception calling "AddTestResult" with "8" argument(s): "The attribute cannot be added because variable Result with value  would no longer be valid."
      at Invoke-Test, /home/chris/.local/share/powershell/Modules/Pester/4.4.3/Functions/It.ps1: line 310
      at ItImpl, /home/chris/.local/share/powershell/Modules/Pester/4.4.3/Functions/It.ps1: line 231
      at It, /home/chris/.local/share/powershell/Modules/Pester/4.4.3/Functions/It.ps1: line 121
      at <ScriptBlock>, /mnt/c/Users/cberg/git/PSScriptAnalyzer/Tests/Documentation/RuleDocumentation.tests.ps1: line 38
      at DescribeImpl, /home/chris/.local/share/powershell/Modules/Pester/4.4.3/Functions/Describe.ps1: line 171
      at Describe, /home/chris/.local/share/powershell/Modules/Pester/4.4.3/Functions/Describe.ps1: line 85
      at <ScriptBlock>, /mnt/c/Users/cberg/git/PSScriptAnalyzer/Tests/Documentation/RuleDocumentation.tests.ps1: line 6
      at <ScriptBlock>, /home/chris/.local/share/powershell/Modules/Pester/4.4.3/Pester.psm1: line 847
      at Invoke-Pester<End>, /home/chris/.local/share/powershell/Modules/Pester/4.4.3/Pester.psm1: line 862
      at <ScriptBlock>, <No file>: line 1
```

**Comment by nohwnd** [comment]
ok, that did not help at all 🤣

**Comment by ffeldhaus** [comment]
I also see this issues showing up in AppVeyor Pester tests after the PowerShell 6.1 update in the recent Ubuntu image update.

**Comment by bergmeister** [comment]
The error comes from [here](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/VariableAttributeCollection.cs#L142) in PowerShell in case that helps as it could be an integration bug with PowerShell as well.

**Comment by ffeldhaus** [comment]
I was able to reproduce it while running Appveyor with SSH access using a the following very simple test:
```powershell
Describe "describe" {
    Context "context" {
        try {
            throw "test"
        }
        catch {
        }
        it "it" {
            throw "inside it"
            Write-Host "everything ok"
        }
    }
}
```

```powershell
PS /home/appveyor> Invoke-Pester -Script "./test.ps1"      
Executing all tests in './test.ps1'

Executing script ./test.ps1

  Describing describe

    Context context
      [-] Error occurred in Context block 62ms
        ValidationMetadataException: The attribute cannot be added because variable Result with value  would no longer be valid.
        MethodInvocationException: Exception calling "AddTestResult" with "8" argument(s): "The attribute cannot be added because variable Result with value  would no longer be valid."
Tests completed in 62ms
Tests Passed: 0, Failed: 1, Skipped: 0, Pending: 0, Inconclusive: 0 
```

This is the PSVersionTable of the AppVeyor image
```powershell
PS /home/appveyor> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      6.1.1
PSEdition                      Core
GitCommitId                    6.1.1
OS                             Linux 4.15.0-43-generic #46~16.04.1-Ubuntu SMP Fri Dec 7 13:31:08 UTC 2018
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

Without the catch block, it does not throw an error. I also confirmed that running the same code on Mac OS X with the same PowerShell version does not result in the issue and instead shows the exception thrown inside the `it` block.

I expect that the Linux PowerShell build somehow has issues with handling exceptions which were captured.

**Comment by ffeldhaus** [comment]
I tracked down the issue to Hashtable properties being case sensitive in the Appveyor Ubuntu (16.04) and Ubuntu1804 (18.04) images. I couldn't recreate the issue with the official Ubuntu docker images and the official steps from Microsoft to install PowerShell.

The following highlights the issue in the Appveyor images:
```powershell
PS /home/appveyor> $test=@{result="result"}
PS /home/appveyor> $test.Result
PS /home/appveyor> $test.result
result
```

Pester defines the [`result` as lowercase property in a Hashtable](https://github.com/pester/Pester/blob/e04f19cfe1f8c7c4bf456fc5b9cc767197a0b809/Functions/Output.ps1#L177) used in the [function `ConvertTo-PesterResult`](https://github.com/pester/Pester/blob/e04f19cfe1f8c7c4bf456fc5b9cc767197a0b809/Functions/Output.ps1#L163). Then the property is accessed as [`$result.Result` starting with an uppercase letter](https://github.com/pester/Pester/blob/d25ff9f34b1fb160df096cc7101efdd8a06cbd43/Functions/It.ps1#L311) in the [`Invoke-Test` function](https://github.com/pester/Pester/blob/d25ff9f34b1fb160df096cc7101efdd8a06cbd43/Functions/It.ps1#L235).

Changing the property access to start with an uppercase would fix the issue. I'd be glad to contribute a pull request if someone can confirm that changing `$result.Result` to `$result.result` is the correct way to fix this. In parallel we can figure out why this is only affecting Appveyor images. [I opened a support ticket to get more information from Appveyor how the build / installed PowerShell](https://help.appveyor.com/discussions/problems/18853-problem-with-powershell-core-611-in-ubuntu-and-ubuntu1804-images).

**Comment by FeodorFitsner** [comment]
I'm not sure the problem is AppVeyor-specific though. These are the results of the same test on my local WSL with Ubuntu 18.04 and PS 6.1.1:

![image](https://user-images.githubusercontent.com/5041459/50541774-8f92b300-0b61-11e9-845c-6cf2dad49256.png)

**Comment by ffeldhaus** [comment]
The issue is very likely caused by https://github.com/PowerShell/PowerShell/issues/7761

**Comment by FeodorFitsner** [comment]
...and HashTable experiment on the same WSL:

![image](https://user-images.githubusercontent.com/5041459/50541786-cd8fd700-0b61-11e9-9eb5-68b1144e7feb.png)

Looks like a regression (or new behavior) in PS Core?

**Comment by bergmeister** [comment]
Today I tried setting the environment variable LANG in AppVeyor to en-us.UTF-8, which did not help, this means that either the environment variable in the yaml gets set too late before pwsh starts or that there is another underlying issue...

**Comment by bergmeister** [comment]
I can confirm that changing the Pester hashtable access in the functions that @ffeldhaus points, it fixes the problem, I opened a PR in Pester since a publishing an updated Version of Pester might be quicker until the proper fix gets deployed in PowerShell 6.2

**Comment by FeodorFitsner** [comment]
Yep, the language should be set "higher" in the process hierarchy. We are working on build agent update. Thanks for your patience guys.

**Comment by nohwnd** [comment]
I just published the referenced PR as 4.4.4 (merged into master and cherrypicked to branch based on 4.4.3). 4.4.4 is the last stable version so AppVeyor users should be getting it unless they specified a version explicitly. Great work guys!

Could someone test it out please ?  @bergmeister ?

**Comment by nohwnd** [comment]
@bergmeister that sucks. Good luck solving it!

**Comment by bergmeister** [comment]
Thanks, I could solve it in the end. I re-ran Pester's test suite on that environment and it only showed me 3 test failures after my PR (before it was 10) and the gist seems to that Pester's mocking feature is not working in those environments (I guess again due to hashtable key casing mismatches either in Pester or PowerShell itself) but I am not using this feature so it is not a problem for me but might be worthwhile to point out to other people in those environments.
For anyone having to fix their product as well, it is worthwhile noting that not only hashtable entries are case sensitive, also doing string comparisons in C# PowerShell modules will suffer from that, the way I solved it was to change from `CurrentCultureIgnoreCase` to `OrdinalIgnoreCase` (note that invariant culture has it's other own issues)
